worker/workertest/workertest.go

```diff
@@ -0,0 +1,315 @@
+package workertest
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http/httptest"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/taskcluster/slugid-go/slugid"
+	tcclient "github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/queue"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+	"github.com/taskcluster/taskcluster-worker/runtime/util"
+	"github.com/taskcluster/taskcluster-worker/worker"
+	"github.com/taskcluster/taskcluster-worker/worker/workertest/fakequeue"
+)
+
+const defaultTestCaseTimeout = 10 * time.Minute
+
+// Case is a worker test case
+type Case struct {
+	Engine            string        // Engine to be used
+	EngineConfig      string        // Engine configuration as JSON
+	PluginConfig      string        // Configuration of plugins, see plugins.PluginManagerConfigSchema()
+	Tasks             []Task        // Tasks to create and associated assertions
+	Concurrency       int           // Worker concurrency, if zero defaulted to 1 and tasks will sequantially dependent
+	StoppedGracefully bool          // True, if worker is expected to stop gracefully
+	StoppedNow        bool          // True, if worker is expected to stop now
+	Timeout           time.Duration // Test timeout, defaults to 10 min
+}
+
+// A Task to be included in a worker test case
+type Task struct {
+	Title           string                  // Optional title (for debugging)
+	Payload         string                  // Task payload as JSON
+	Success         bool                    // True, if task should be successfully
+	Exception       runtime.ExceptionReason // Reason, if exception is expected
+	Artifacts       ArtifactAssertions      // Mapping from artifact name to assertion
+	AllowAdditional bool                    // True, if additional artifacts is allowed
+	Status          StatusAssertion         // Optional, custom assertion on status and queue
```

Comment:
I don't expect we'll use this much, if ever... it's here to allow any custom checks you might want to do.

The signature is `func(t *testing.T, q *queue.Queue, status queue.TaskStatusStructure)` so you can fetch artifacts from a run that isn't the latest, if you wish to do that...